### ISSUE TITLE
Adding JAX support for FlexFlow

### ISSUE THREAD

**Issue by lockshaw** [flexflow/FlexFlow#798]
Moved from #483 

1. > @rajasbansal 
   > There are two issues with trying to implement JAX in the FlexFlow environment -
   > 
   > **1. Level of tracing in JAXPR**
   > 
   > JAX is traced at a low level representation called JAXPr. Vanilla JAX does not contain higher order operations like Linear, LSTM etc out of the box and there are other libraries like [Haiku](https://github.com/deepmind/dm-haiku) and [Flax](https://github.com/google/flax) which sit on top of JAX and provide these abstractions to help people build deep learning models.
   > 
   > There is no higher level representation which would help us, unless we choose one library and do the translation at the library level. 
   > 
   > A simple example of JAXPr translation is attached below.
   > ```
   > class MLP():
   >   def __init__(self):
   >     pass
   > 
   >   def forward(self, params, x):
   >     w, b = params
   >     return jax.nn.softmax((jnp.matmul(w, x) + b))
   > ```
   > This is a simple JAX model for defining one MLP layer over an input.
   > ```
   > { lambda ; a:f32[5,3] b:f32[5,1] c:f32[3,1]. let
   >     d:f32[5,1] = dot_general[
   >       dimension_numbers=(((1,), (0,)), ((), ()))
   >       precision=None
   >       preferred_element_type=None
   >     ] a c
   >     e:f32[5,1] = add d b
   >     f:f32[5] = reduce_max[axes=(1,)] e
   >     g:f32[5,1] = broadcast_in_dim[broadcast_dimensions=(0,) shape=(5, 1)] f
   >     h:f32[5,1] = stop_gradient g
   >     i:f32[5,1] = sub e h
   >     j:f32[5,1] = exp i
   >     k:f32[5] = reduce_sum[axes=(1,)] j
   >     l:f32[5,1] = broadcast_in_dim[broadcast_dimensions=(0,) shape=(5, 1)] k
   >     m:f32[5,1] = div j l
   >   in (m,) }
   > ```
   > This is the JAXPr translation of the MLP layer. Here we see that there is no way to figure out that there is a linear layer or there is a softmax operation, both of which are optimized in the flexflow model.
   > 
   > There are 3 options to tackle this problem
   > 1.1 Write a higher level translator in Python which works on the JAXPr and tries to infer higher level operations by combining lower level operations into one single operation.
   > 1.2 Combine operations as seen by the FlexFlow layer and fuse them together. Here the optimization of the model is shipped to the FlexFlow layer.
   > 1.3 Choose one deep learning library (Haiku) which looks promising and develop the translator for that library. In this case, we would also have to write our tracer which traces the models being used (like torch.fx).
   > 
   > **2. Defining the parameters for taking gradients** 
   > 
   > JAX requires all functions to be pure, which means that parameters of the model are also sent as an input to the graph of operations which is being traced. Looking at the JAX representation of the MLP, you can see that the forward function has the model parameters as an argument to the function. However, FlexFlow gives the ownership of the model weights to the operators, with the operators allocating their own weight tensor.
   > 
   > Then, for allowing backpropagation and updation of the parameters of the model, how will these be handled by FlexFlow. Will they just be sent to the operator separately as tensors which are parameters of the model and thus should be updated at every step?
   > 
   > An alternative approach would be to allow all operators to have the option of passing the weight tensor as an argument instead of allocating their own weight tensor. However, this approach has the problem that there might be some parameters which JAX requires gradients for, but are not owned by any operator. For example, if jax is trying to learn W in the function `f(x) = Wx`, then the batch multiplication operator does not own W, however, it is a parameter which needs to be updated based on gradients.
   > 
   > @lockshaw @jiazhihao @eddy16112 Any thoughts on what approach seems right to tackle these issues?
 
2. > @lockshaw 
   > 1. I think 1.1 is the correct approach for now (potentially not the final solution, but the one we should start with). Solution 1.3 seems likely to doom us to endless engineering, so I think we can treat that as a worst case scenario, as I assume there is no small amount of library surface area to cover. I think 1.2 is probably possible and _might_ have some benefits in the long run, but would require not so much a lot of engineering but a lot of thinking/algorithms/development, which we should also avoid if there is an easier solution. In the absence of any other information, I'd say we have @rajas1998 progress along the lines of 1.1, and @rajas1998 can keep us up-to-date on how it is going and we can pivot to a different solution later once we actually understand the problems that prevent 1.1.
   > 
   > 2. I think the correct solution here is to allow operators to take in weights rather than own them, and directly materialize the weight nodes in the PCG as we have been meaning to do for a while (I think this would also ease using the same parameters across different operators--I think this might be done in the transformer encoder/decoding layers though I could be wrong--which I doubt we currently handle). I think the best way to handle this is to start prototyping out a refactored operator interface (ideally once the kernel refactors (i.e., #303) are done so that the person doing this isn't waiting for long builds constantly. 
   > 
   > I guess in summation: I think in the case of both problems we should do some prototyping on a small scale and then evaluate rather than trying to make a final decision now. Following this assumption, the question becomes @jiazhihao @eddy16112 do you agree with these initial prototyping directions, or do you think a different approach would be better to start with?
   
3. > @lockshaw 
   > @jiazhihao @eddy16112 Thoughts?
   
4. > @eddy16112
   > If I understand 1.1 correctly, we are trying to take the JAXPr, identify the low level operations and combine them into high level FlexFlow operators, right? Given the JAXPr of MLP, I just feel like it is difficulty to tell these operators can be combined into a linear Op.
